### ISSUE TITLE
iwyu asserts on include cycle

### ISSUE THREAD

**Issue by zjturner** [include-what-you-use/include-what-you-use#424]
Shouldn't this just output a diagnostic like everything else?  I mean it's a legitimate coding error that someone can make which would not cause a compile error (which is indeed how I found this assertion, I had a header file which inadvertently included itself).

```
// Foo.h
#ifndef FOO_H
#define FOO_H
#include "Foo.h"  // iwyu asserts, but any compiler will happily accept this.
#endif
```

**Comment by kimgr** [comment]
Boost has this a lot too, apparently, see #406.

I haven't dug into the details yet, but I can imagine the cycle will throw IWYU's analysis for a fit, since it sometimes needs to follow the include graph.

**Comment by zjturner** [comment]
Even if it follows the include graph, the preprocessor should magically
handle it the second time around, just like in the real program (ie when
you do the second include, the include guard is defined, so the whole file
evaluates to nothing)

**Comment by vsapsai** [comment]
@zjturner, can you reproduce the problem with exact code you've provided? Because we try to handle at least simple cycles (see [tests/cxx/include_cycle.cc](https://github.com/include-what-you-use/include-what-you-use/blob/master/tests/cxx/include_cycle.cc)). I am wondering if there is something else that triggers this problem.

**Comment by kimgr** [comment]
I can't reproduce this. Tried with a `Foo.cc` file including your `Foo.h` above, a `Foo_main.cc` including `Foo.h` to test when they're not associated, and running IWYU on `Foo.h` directly. No assert.

I did see some annoying absolute path suggestions, though, but it feels like you're seeing something we're not. I can see if I can get my environment up and running on Windows, and try there, in case this is something Windows-specific.

**Comment by zjturner** [comment]
I encountered this when building llvm / lldb, so you might try to just run
against llvm on whatever platform and see if it repros. If you use that
magic cmake define it makes it really easy

**Comment by banjo74** [comment]
Our organization uses -isystem for third party software (including boost).  We do this to also ignore compiler warnings that arise from such headers.  Maybe iwyu could ignore such issues that are included via -isystem.

**Comment by dpelle** [comment]
I encountered the assert failure too. In case it helps, here is what I see:

{noformat}
Cycle in include-mapping:
  <boost/phoenix/support/detail/iterate.hpp> ->
  <boost/phoenix/operator/detail/cpp03/mem_fun_ptr_eval_result_of.hpp> ->
  <boost/phoenix/operator/detail/cpp03/mem_fun_ptr_eval_result_of.hpp>
/mnt/buildbot_iwyu_trunk/workspace/release/3.9/include-what-you-use/iwyu_include_picker.cc:844: Assertion failed: Cycle in include-mapping
Aborted (core dumped)
{noformat}

Another occurrence:
{noformat}
Cycle in include-mapping:
  <boost/variant/detail/substitute.hpp> ->
  <boost/preprocessor/iteration/detail/iter/forward1.hpp> ->
  <boost/preprocessor/iteration/detail/iter/forward1.hpp>
/mnt/buildbot_iwyu_trunk/workspace/release/3.9/include-what-you-use/iwyu_include_picker.cc:844: Assertion failed: Cycle in include-mapping
Aborted (core dumped)
{noformat}

**Comment by kimgr** [comment]
@dpelle: thanks! Can you also share your `include-what-you-use` invocation that triggers this (preferably with a reduced test case)?

**Comment by mblythe86** [comment]
I've hit the error with boost as well, and while I don't have a reduced testcase, I can lightswitch the problem by switching between boost-1.61 and boost-1.62.

On failure, I see this when I turn up verbosity:

> Found include: /path/to/boost-1.62/include/boost/variant/detail/substitute.hpp:162 -> /path/to/boost-1.62/include/boost/preprocessor/iteration/detail/iter/forward1.hpp
> Found include: /path/to/boost-1.62/include/boost/preprocessor/iteration/detail/iter/forward1.hpp:17 -> /path/to/boost-1.62/include/boost/preprocessor/iteration/detail/bounds/lower1.hpp
> Found include: /path/to/boost-1.62/include/boost/preprocessor/iteration/detail/iter/forward1.hpp:19 -> /path/to/boost-1.62/include/boost/preprocessor/iteration/detail/bounds/upper1.hpp
> Found include: /path/to/boost-1.62/include/boost/preprocessor/iteration/detail/iter/forward1.hpp:47 -> /path/to/boost-1.62/include/boost/variant/detail/substitute.hpp
> Found include: /path/to/boost-1.62/include/boost/preprocessor/iteration/detail/iter/forward1.hpp:52 -> /path/to/boost-1.62/include/boost/variant/detail/substitute.hpp
> Found include: /path/to/boost-1.62/include/boost/preprocessor/iteration/detail/iter/forward1.hpp:57 -> /path/to/boost-1.62/include/boost/variant/detail/substitute.hpp
> Found include: /path/to/boost-1.62/include/boost/preprocessor/iteration/detail/iter/forward1.hpp:62 -> /path/to/boost-1.62/include/boost/variant/detail/substitute.hpp
> Found include: /path/to/boost-1.62/include/boost/preprocessor/iteration/detail/iter/forward1.hpp:67 -> /path/to/boost-1.62/include/boost/variant/detail/substitute.hpp
> Found include: /path/to/boost-1.62/include/boost/preprocessor/iteration/detail/iter/forward1.hpp:72 -> /path/to/boost-1.62/include/boost/variant/detail/substitute.hpp
> Adding mapping from <boost/variant/detail/substitute.hpp> to <boost/preprocessor/iteration/detail/iter/forward1.hpp>
> Adding mapping from <boost/preprocessor/iteration/detail/iter/forward1.hpp> to <boost/variant/detail/substitute.hpp>
> 

On pass, I see this:

> Found include: /path/to/boost-1.61/include/boost/variant/detail/substitute.hpp:162 -> /path/to/boost-1.61/include/boost/preprocessor/iteration/detail/iter/forward1.hpp
> Found include: /path/to/boost-1.61/include/boost/preprocessor/iteration/detail/iter/forward1.hpp:17 -> /path/to/boost-1.61/include/boost/preprocessor/iteration/detail/bounds/lower1.hpp
> Found include: /path/to/boost-1.61/include/boost/preprocessor/iteration/detail/iter/forward1.hpp:19 -> /path/to/boost-1.61/include/boost/preprocessor/iteration/detail/bounds/upper1.hpp
> Found include: /path/to/boost-1.61/include/boost/preprocessor/iteration/detail/iter/forward1.hpp:47 -> /path/to/boost-1.61/include/boost/variant/detail/substitute.hpp
> Found include: /path/to/boost-1.61/include/boost/preprocessor/iteration/detail/iter/forward1.hpp:52 -> /path/to/boost-1.61/include/boost/variant/detail/substitute.hpp
> Found include: /path/to/boost-1.61/include/boost/preprocessor/iteration/detail/iter/forward1.hpp:57 -> /path/to/boost-1.61/include/boost/variant/detail/substitute.hpp
> Found include: /path/to/boost-1.61/include/boost/preprocessor/iteration/detail/iter/forward1.hpp:62 -> /path/to/boost-1.61/include/boost/variant/detail/substitute.hpp
> Found include: /path/to/boost-1.61/include/boost/preprocessor/iteration/detail/iter/forward1.hpp:67 -> /path/to/boost-1.61/include/boost/variant/detail/substitute.hpp
> Found include: /path/to/boost-1.61/include/boost/preprocessor/iteration/detail/iter/forward1.hpp:72 -> /path/to/boost-1.61/include/boost/variant/detail/substitute.hpp
> Adding mapping from <boost/preprocessor/iteration/detail/iter/forward1.hpp> to "boost/variant/detail/substitute.hpp"
> Adding mapping from <boost/variant/detail/substitute.hpp> to <boost/preprocessor/iteration/detail/iter/forward1.hpp>
> 

The difference in quoting seems to make all the difference, and I can see that the quoting did, in fact, change between the boost releases: https://github.com/boostorg/variant/compare/boost-1.61.0...boost-1.62.0#diff-931166bdb0e6db843850b879dce8ea83

Unfortunately for IWYU, I don't think boost is doing anything incorrect here...just that their "preprocessor iterate" stuff is really gross.  I agree with what banjo74 said...IWYU should probably just ignore errors like this that happen in system headers.

**Comment by kimgr** [comment]
OK, thanks! Looks like something missteps when the same header is included with both `<>` and `""`` *and* is in a cycle.

**Comment by zjturner** [comment]
I'd be a bit surprised if the original report is the same, because we never
#include via <> except system includes. But maybe there's an odd <> in
there somewhere

**Comment by kimgr** [comment]
I jumped to the wrong conclusion in https://github.com/include-what-you-use/include-what-you-use/issues/424#issuecomment-305974359 -- if anything, @mblythe86's comment indicates that consistent use of `<>` would be problematic.

But there's something more going on -- the `<boost/variant.hpp>` case only fails when `-isystem` is used. Using `-isystem` on a smaller case does not exhibit the problem. :-/

I'll keep digging, but I'm pretty sure this is related to #406.

**Comment by larseggert** [comment]
FWIW, I hit this bug just know on the `Uri.h` file of http://uriparser.sourceforge.net/

**Comment by kimgr** [comment]
@larseggert - can you also share the `include-what-you-use` invocation that makes this happen? If you're using `iwyu_tool.py`, I believe the `-v` switch prints it to screen.

**Comment by DaveTrussel** [comment]
In case it is of any help. Came across the same bug when using `boost_python`. Invoked `iwyu` with CMake `set(CMAKE_CXX_INCLUDE_WHAT_YOU_USE ${iwyu_path} -Xiwyu --no_comments)`

Includes in cpp file:
```
#include <boost/python.hpp>
#include <boost/python/suite/indexing/vector_indexing_suite.hpp>
```
Error:
```
Cycle in include-mapping:
  <boost/python/signature.hpp> ->
  <boost/preprocessor/iteration/detail/iter/forward2.hpp> ->
  <boost/python/signature.hpp> ->
  <boost/python/signature.hpp>
/home/dave/llvm-trunk/llvm/tools/clang/tools/include-what-you-use/iwyu_include_picker.cc:845: Assertion failed: Cycle in include-mapping
Child aborted

```
`include-what-you-use 0.9 (git:d3e8a0d) based on clang version 6.0.0 (trunk 321461)`

**Comment by xlicz** [comment]
+1. this problem also happen to my project, don't know why, I use include-what-you-use follow this thread:[https://stackoverflow.com/questions/30951492/how-to-use-the-tool-include-what-you-use-together-with-cmake-to-detect-unused-he](url)

**Comment by JIghtuse** [comment]
Here is a minimal example which triggers cycle for me:

```
#include <boost/variant.hpp>

int main()
{
}
```

I found that it can be fixed when I use separately headers with definitions for `boost::variant` type and, say, `boost::get`:
```
#include <boost/variant/get.hpp>
#include <boost/variant/variant.hpp>

int main()
{
}
```

```
user@host:p$ /usr/bin/iwyu  /tmp/v.cpp

/tmp/v.cpp should add these lines:

/tmp/v.cpp should remove these lines:
- #include <boost/variant/get.hpp>  // lines 1-1
- #include <boost/variant/variant.hpp>  // lines 2-2

The full include-list for /tmp/v.cpp:
---
```

Logs when it cycles:
```
user@host:p$ /usr/bin/iwyu  /tmp/v.cpp
Cycle in include-mapping:
  <boost/variant/detail/substitute.hpp> ->
  <boost/preprocessor/iteration/detail/iter/forward1.hpp> ->
  <boost/preprocessor/iteration/detail/iter/forward1.hpp>
/build/iwyu-AFSZiE/iwyu-4.0/iwyu_include_picker.cc:845: Assertion failed: Cycle in include-mapping
Aborted (core dumped)
user@host:p$ /usr/bin/iwyu --verbose /tmp/v.cpp
clang version 4.0.1-1build1 (tags/RELEASE_401/final)
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
Found candidate GCC installation: /usr/bin/../lib/gcc/i686-linux-gnu/7
Found candidate GCC installation: /usr/bin/../lib/gcc/i686-linux-gnu/7.2.0
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/4.9
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/4.9.3
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/5
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/5.4.1
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/6
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/6.4.0
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/7
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/7.2.0
Found candidate GCC installation: /usr/lib/gcc/i686-linux-gnu/7
Found candidate GCC installation: /usr/lib/gcc/i686-linux-gnu/7.2.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/4.9
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/4.9.3
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/5
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/5.4.1
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/6.4.0
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/7.2.0
Selected GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/7.2.0
Candidate multilib: .;@m64
Selected multilib: .;@m64
clang invocation:
 "/usr/bin/include-what-you-use" "-cc1" "-triple" "x86_64-pc-linux-gnu" "-fsyntax-only" "-disable-free" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "v.cpp" "-mrelocation-model" "static" "-mthread-model" "posix" "-mdisable-fp-elim" "-fmath-errno" "-masm-verbose" "-mconstructor-aliases" "-munwind-tables" "-fuse-init-array" "-target-cpu" "x86-64" "-v" "-dwarf-column-info" "-debugger-tuning=gdb" "-resource-dir" "/usr/bin/../lib/clang/4.0.1" "-internal-isystem" "/usr/bin/../lib/gcc/x86_64-linux-gnu/7.2.0/../../../../include/c++/7.2.0" "-internal-isystem" "/usr/bin/../lib/gcc/x86_64-linux-gnu/7.2.0/../../../../include/x86_64-linux-gnu/c++/7.2.0" "-internal-isystem" "/usr/bin/../lib/gcc/x86_64-linux-gnu/7.2.0/../../../../include/x86_64-linux-gnu/c++/7.2.0" "-internal-isystem" "/usr/bin/../lib/gcc/x86_64-linux-gnu/7.2.0/../../../../include/c++/7.2.0/backward" "-internal-isystem" "/usr/include/clang/4.0.1/include/" "-internal-isystem" "/usr/local/include" "-internal-isystem" "/usr/bin/../lib/clang/4.0.1/include" "-internal-externc-isystem" "/usr/include/x86_64-linux-gnu" "-internal-externc-isystem" "/include" "-internal-externc-isystem" "/usr/include" "-fdeprecated-macro" "-fdebug-compilation-dir" "/home/user/projects/p" "-ferror-limit" "19" "-fmessage-length" "172" "-fobjc-runtime=gcc" "-fcxx-exceptions" "-fexceptions" "-fdiagnostics-show-option" "-x" "c++" "/tmp/v.cpp"

clang -cc1 version 4.0.1 based upon LLVM 4.0.1 default target x86_64-pc-linux-gnu
ignoring nonexistent directory "/include"
ignoring duplicate directory "/usr/bin/../lib/gcc/x86_64-linux-gnu/7.2.0/../../../../include/x86_64-linux-gnu/c++/7.2.0"
ignoring duplicate directory "/usr/include/clang/4.0.1/include"
#include "..." search starts here:
#include <...> search starts here:
 /usr/bin/../lib/gcc/x86_64-linux-gnu/7.2.0/../../../../include/c++/7.2.0
 /usr/bin/../lib/gcc/x86_64-linux-gnu/7.2.0/../../../../include/x86_64-linux-gnu/c++/7.2.0
 /usr/bin/../lib/gcc/x86_64-linux-gnu/7.2.0/../../../../include/c++/7.2.0/backward
 /usr/include/clang/4.0.1/include
 /usr/local/include
 /usr/include/x86_64-linux-gnu
 /usr/include
End of search list.
Cycle in include-mapping:
  <boost/variant/detail/substitute.hpp> ->
  <boost/preprocessor/iteration/detail/iter/forward1.hpp> ->
  <boost/preprocessor/iteration/detail/iter/forward1.hpp>
/build/iwyu-AFSZiE/iwyu-4.0/iwyu_include_picker.cc:845: Assertion failed: Cycle in include-mapping
Aborted (core dumped)
```

Boost version: 1.62.0.1
Distro: Ubuntu 17.10

**Comment by wthrowe** [comment]
Here's a minified test case (for iwyu 0.9, haven't tested later):
```
$ cat main.cpp
#include <z.hpp> // works with ""
$ cat z.hpp
#ifndef z_hpp
#define z_hpp
#include <z.hpp> // "" vs. <> doesn't matter
#endif
$ include-what-you-use -isystem . main.cpp
Cycle in include-mapping:
  <z.hpp> ->
  <z.hpp>
/var/tmp/portage/dev-util/include-what-you-use-5.0/work/include-what-you-use-clang_5.0/iwyu_include_picker.cc:845: Assertion failed: Cycle in include-mapping
Aborted
$ include-what-you-use --version
include-what-you-use 0.9 based on clang version 5.0.1 (tags/RELEASE_501/final)
```
As noted in the comments, changing `<>` to `""` in the main file works around the failure.

My analysis is that there are actually two bugs triggering here: iwyu can't handle include cycles (this bug), and iwyu can't reliably detect cycles (causing `<>` vs. `""` to matter).  The second is due to iwyu having two ways of deciding which kind of quoting to use.  Sometimes it decides based on whether the file is in a system include directory, and sometimes it uses whichever quoting style it first encountered for the file.  For the above test case both methods agree and a `<z.hpp> -> <z.hpp>` cycle is generated, but if the include from `min.cpp` is changed to `"z.hpp"` then instead `<z.hpp> -> "z.hpp"` is generated, which is not recognized as a loop.  (Note that the quoting there is the reverse of the naive guess: the file reached through a `""` include is named with `<>` and vice versa.)

As for this bug, the failure is triggered while attempting to compute the transitive closure of the include graph (as explained in the source code).  It is not immediately clear to me whether the failure simply indicates that the transitive closure algorithm has not been coded to handle cycles, or whether something later is assuming the graph is acyclic and this is just the easiest place to check that.  I can write a patch to handle cycles in the closure algorithm if that is useful.

**Comment by Archiatrus** [comment]
Is there any work around to this problem? I would like to use it (version 0.10 for clang-6.0) with `-DCMAKE_CXX_INCLUDE_WHAT_YOU_USE="/usr/bin/iwyu;-Xiwyu;--no_fwd_decls"`. But I get
```
Cycle in include-mapping:
  <boost/utility/in_place_factory.hpp> ->
  <boost/preprocessor/iteration/detail/iter/forward1.hpp> ->
  <boost/preprocessor/iteration/detail/iter/forward1.hpp>
/home/IBEO.AS/dako/Documents/include-what-you-use/iwyu_include_picker.cc:843: Assertion failed: Cycle in include-mapping
Child aborted
```
Changing the boost version is not an option (or changing <> to "" in the boost headers).

**Comment by Archiatrus** [comment]
Ok I "fixed" this by adding "|| key.find("boost/preprocessor") != string::npos" in iwyu_include_picker.cc line 832. By default it only ignores third party code that is included with `third_party/something`.

**Comment by kimgr** [comment]
It just occurred to me (after 2.5 years of thinking about this off and on) that this checks for a cycle *in mappings*, not in the *include graph*.

But I'm guessing none of you run with the provided Boost mappings, so I'm going to assume this is somehow caused by IWYU's auto-mapping (there are a few places where it adds mapping based on qualities of the include graph). So I suspect maybe we should break the cycle when constructing the mappings, not when post-analyzing them.

**Comment by thehans** [comment]
I have a cycle issue with this library header which includes itself:
https://github.com/CGAL/cgal/blob/master/Number_types/include/CGAL/Gmpq.h#L187

I don't completely understand the comment `//since types are included by Gmp_coercion_traits.h:` if that explains some reason it is needed, or if its just an extraneous line.  As far as I can tell the line shouldn't do anything since its within its own #ifdef block.

Is there a way to workaround with iwyu besides editing this library's header?

**Comment by kimgr** [comment]
@thehans I recommend running IWYU with a higher verbosity to try and understand what decision is made that leads up to this cycle. Please try adding `-Xiwyu -v7` to the IWYU command, and see if you can dig something out of the flow.

**Comment by thehans** [comment]
@kimgr I used `-v7` and the output is like 16k lines for the first c++ file that fails.

Besides `Gmpq.h` directly including itself, it includes a bunch of other files which include `Gmpq.h` and they include files which also include `Gmpq.h` etc.
There are too many cycles to count by hand, but I tried to detail overall the situation by grepping the output many times:
[Cycles.txt](https://github.com/include-what-you-use/include-what-you-use/files/3856178/Cycles.txt)

Not sure if that is of any use.  Can I just exclude iwyu from looking down this tree of includes?

**Comment by devbharat** [comment]
I also have this issue with clang-6.0 and boost1.62

```
Error running '/usr/local/bin/include-what-you-use': Cycle in include-mapping:
  <boost/phoenix/support/detail/iterate.hpp> ->
  <boost/phoenix/operator/detail/cpp03/mem_fun_ptr_eval_result_of.hpp> ->
  <boost/phoenix/operator/detail/cpp03/mem_fun_ptr_eval_result_of.hpp>
/home/bharat/Documents/Tools/include-what-you-use/iwyu_include_picker.cc:843: Assertion failed: Cycle in include-mapping

```

**Comment by Archiatrus** [comment]
Have a look at the provided mapping files. Boost for example has tons of cyclic includes and with the provided mapping files you can get around them (you maybe need to create your own but they tell you how). From the examples it should be possible to figure out how to break cyclic includes in other libs... Except maybe the self include.

**Comment by devbharat** [comment]
I have not! Could briefly explain the process or link to a doc?

**Comment by Archiatrus** [comment]
See the [Docs](https://github.com/include-what-you-use/include-what-you-use/blob/master/docs/IWYUMappings.md) and the mapping files you find in this repo. For example boost-all-private.imp.

**Comment by srpgilles** [comment]
Is this ticket still active? (not very hopeful looking at #845...)

 I've tried to use include-what-you use following the [incentive](https://www.fluentcpp.com/2021/01/01/include-what-you-use/) to do so given on [FluentCpp](https://www.fluentcpp.com), but I'm stuck with this issue with a third party library I use ([Xtensor-blas](https://github.com/xtensor-stack/xtensor-blas)) and it's not obvious to me how to circumvent it.

Thanks!

**Comment by dpelle** [comment]
For the record, I see this kind of assertion with include cycle:
```
Cycle in include-mapping:
  <boost/spirit/home/support/detail/as_variant.hpp> ->
  <boost/preprocessor/iteration/detail/iter/forward1.hpp> ->
  <boost/preprocessor/iteration/detail/iter/forward1.hpp>
/home/dope/sb/include-what-you-use/iwyu_include_picker.cc:932: Assertion failed: Cycle in include-mapping                                                                                                                                                                        
Aborted (core dumped)
```

My version of include-what-you-use:
```
$ include-what-you-use  --version
include-what-you-use 0.15 (git:5db414a) based on Ubuntu clang version 11.0.1-++20201218093139+43ff75f2c3fe-1~exp1~20201218203800.157
```

I've seen those assert for a long time since I reported the same issue in Apr 2018: https://github.com/include-what-you-use/include-what-you-use/issues/534

**Comment by srpgilles** [comment]
@dpelle If you're using it since at least 2018, can you tell how you cope in practice with this assert?

I haven't yet found a way to circumvent it for my code.

Thanks!

**Comment by srpgilles** [comment]
I finally delved into Xtensor-blas code, and the issue was in my case straightforward (two files were included into themselves).

Still, it would be nice if iwyu could be more helpful and provide a way to circumvent the issue in the most problematic cases - especially when the problem stems from third party code.

**Comment by dpelle** [comment]
@srpgilles asked:

> @dpelle If you're using it since at least 2018, can you tell how you cope in practice with this assert?

Fortunately, it does not happen on all the source files that I check with iwyu.
When the assert happens in a source file, I simply ignore iwyu output for that source file.
It's not course not ideal, but since iwyu works for most files, iwyu is still useful.

Some stats on the project where I used iwyu:
*2636* files give the assert failure "Assertion failed: Cycle in include-mapping" out of *9304* files checked. 

Interestingly, when I noticed this a long time ago, it was happening in boost header. But now I see
that all occurrences are in the same include which is part of my application.  Hmm, I need to check
this and perhaps I can change my code to avoid it.

**Comment by dpelle** [comment]
@dpelle wrote:

> Interestingly, when I noticed this a long time ago, it was happening in boost header. But now I see
> that all occurrences are in the same include which is part of my application. Hmm, I need to check
> this and perhaps I can change my code to avoid it.

I think I fixed the recursive include in my project. Somehow it compiled fine with gcc or clang,
however, at least include-what-you-use asserted and clangd (which I used t jump to definition
via YouCompleteMe vim plugin) was giving an error.  I'll ran include-what-you-use over night.
I'm hoping I won't run into that assert in iwyu anymore in my project.  Still, ideally include-what-you-use
should cope with this (possibly give a warning) as clang or gcc don't seem to bother.

**Comment by petersteneteg** [comment]
I ran into the same issue with Qt, qglobal.h has some cycle,. I just added an test after this
https://github.com/include-what-you-use/include-what-you-use/blob/1d053edce2169209ad4be8fbf3031eec414ffef2/iwyu_include_picker.cc#L1098-L1101
to also ignore qglobal.h seems to work fine for qt. 
The current implementations behavior of having hardcoded an ignore on all "internal/*" headers seems really odd, and awfully project specific. One would which that all such behaviors were configurable somewhere and not hardcoded.

**Comment by kimgr** [comment]
While trying to get my head around consequences of #1137, I stumbled on:
https://github.com/include-what-you-use/include-what-you-use/issues/424#issuecomment-377722731. After having looked into the peculiarities of implicit mappings and cycle detection, I can only agree with @wthrowe's analysis there.

As far as I can make out, the transitive closure algorithm as-is can handle/break cycles, but I'm not entirely sure if the resulting mappings are representative of a cyclic input include graph.

**Comment by kimgr** [comment]
I've played with this a little bit, and one issue came in with [this patch](https://github.com/include-what-you-use/include-what-you-use/commit/04f1c925370d0f1a70642147f07c70a4c2bfb530) in August 2016.

It added some mechanics for tracking "outward" macro dependencies, and recorded them as automatic mappings. This triggered for include guards between e.g. `"z.hpp"` and `<z.hpp>`, and would add a direct self-mapping.

That patch has been in place since IWYU 0.7, which matches the timeline for when these problems started to occur.

I'll cook up a patch to avoid direct self-mapping, but I suspect that won't be enough for the boost use-case, where there seems to be longer cycles.

**Comment by kimgr** [comment]
Here's a first attempt: https://github.com/include-what-you-use/include-what-you-use/pull/1150.

I think that should take us back to the state before IWYU 0.7, but I don't think it will swallow all of Boost.

**Comment by kimgr** [comment]
After some digging around, I can't convince myself that cycles in mappings are harmful at all for IWYU's analysis. As far as I can tell, the only time it traverses the mapping table as a graph is when computing the transitive closure.

Ignoring cycles there should be benign as long as we visit all the nodes in DFS order. The only difference for nodes with cycles is that they will not be traversed as part of child traversal. This, as far as I can tell, does not affect their transitive closure.

For example, here's a possible mapping table with cycles:
```
<a.h>: [<b.h>, <c.h>]
<b.h>: [<d.h>]
<c.h>: [<a.h>]
```
Once the closure is completed, the table should look something like this:
```
<a.h>: [<b.h>, <c.h>, <d.h>]  # <a.h> via <c.h> was ignored because of cycle
<b.h>: [<d.h>]
<c.h>: [<a.h>, <b.h>, <d.h>]  # <c.h> via <a.h> was ignored because of cycle
```

Now, unless I'm missing something, when mapping happens, the only thing we do is a simple lookup of all public headers for e.g. `<a.h>`, which would then give a vector of `[<b.h>, <c.h>, <d.h>]`. Then we attempt to select the best candidate from that list. There is no recursive mapping, as far as I can tell, since that has already been expanded by the transitive closure.

I think that also means the `""` vs `<>` cycle detection is not critical. As we saw above, the cycle detection is only necessary to protect the transitive closure algorithm from going into infinite recursion, and if `"a.h" -> <a.h>` is not a cycle, it will not exhibit poor behavior. It might put a redundant/bad `"a.h" -> <a.h>` mapping in the final mapping table, but that's a separate problem (captured by the [angle-quote-curse](https://github.com/include-what-you-use/include-what-you-use/labels/angle-quote-curse) issue label)

**Comment by kimgr** [comment]
Hey all, sorry about the extremely long wait for this trivial fix. I just needed to think through the whole problem :)

I would be extremely grateful for some testing feedback (or review feedback, for that matter) on this PR: https://github.com/include-what-you-use/include-what-you-use/pull/1152.

**Comment by kimgr** [comment]
Presumably fixed by PR #1152
